Cannot convert undefined or null to object at `@pulumi/runtime/closure/codePaths.ts:250:34`

### What happened?

I was trying to run the default boilerplate generated by `pulumi new serverless-aws-typescript` and was faced with the following error:

```sh
$ pulumi up
Previewing update ([REDACTED])

View in Browser (Ctrl+O): https://app.pulumi.com/[REDACTED]/previews/57a4758b-7849-41ce-bdf2-bb4e349ade9c

     Type                 Name                Plan     Info
     pulumi:pulumi:Stack  [REDACTED]                   1 error; 2 messages

Diagnostics:
  pulumi:pulumi:Stack ([REDACTED]):
    (node:33492) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
    (Use `node --trace-deprecation ...` to show where the warning was created)

    error: Running program '[REDACTED]/packages/infrastructure/' failed with an unhandled exception:
    <ref *1> TypeError: Cannot convert undefined or null to object
        at Function.keys (<anonymous>)
        at [REDACTED]/node_modules/@pulumi/runtime/closure/codePaths.ts:250:34
        at Generator.next (<anonymous>)
        at [REDACTED]/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:21:71
        at new Promise (<anonymous>)
        at __awaiter ([REDACTED]/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:17:12)
        at allFoldersForPackages ([REDACTED]/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:166:12)
        at [REDACTED]/node_modules/@pulumi/runtime/closure/codePaths.ts:118:37
        at Generator.next (<anonymous>)
        at [REDACTED]/node_modules/@pulumi/pulumi/runtime/closure/codePaths.js:21:71 {
      promise: Promise { <rejected> [Circular *1] }
    }


npm ERR! Lifecycle script `dev` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: [REDACTED] 
npm ERR!   at location: [REDACTED]/packages/infrastructure
```

### Example

Here are the steps I followed:

1. Create an NPM workspace
2. Create a workspace project at `packages/infrastructure`
3. Run `pulumi new serverless-aws-typescript`
4. Move dependencies from `packages/infrastructure/package.json` to the root `package.json`
5. Run `pulumi up` in `packages/infrastructure`

### Output of `pulumi about`

```sh
$ pulumi about
CLI          
Version      3.104.2
Go Version   go1.21.6
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host     
OS       darwin
Version  13.6.6
Arch     arm64

This project is written in nodejs: executable='/Users/benheidemann/.nvm/versions/node/v22.0.0/bin/node' version='v22.0.0'

Current Stack: kmbal/kmbal-mono/staging

TYPE                                               URN
pulumi:pulumi:Stack                                urn:pulumi:staging::kmbal-mono::pulumi:pulumi:Stack::kmbal-mono-staging
pulumi:providers:aws                               urn:pulumi:staging::kmbal-mono::pulumi:providers:aws::default_6_18_2
aws:iam/role:Role                                  urn:pulumi:staging::kmbal-mono::aws:iam/role:Role::fn
aws:lambda/function:Function                       urn:pulumi:staging::kmbal-mono::aws:lambda/function:Function::fn
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-019020e7
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-7cd09230
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-e1a3786d
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-1b4caae3
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-74d12784
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-d32a66fa
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-4aaabb8e
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-b5aeb6b6
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws:iam/rolePolicyAttachment:RolePolicyAttachment::fn-a1de8170
pulumi:providers:aws-apigateway                    urn:pulumi:staging::kmbal-mono::pulumi:providers:aws-apigateway::default_2_1_1
pulumi:providers:pulumi                            urn:pulumi:staging::kmbal-mono::pulumi:providers:pulumi::default
aws-apigateway:index:RestAPI                       urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI::api
pulumi:providers:aws                               urn:pulumi:staging::kmbal-mono::pulumi:providers:aws::default_6_0_4
aws:iam/role:Role                                  urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:iam/role:Role::api4c238266
aws:s3/bucket:Bucket                               urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:s3/bucket:Bucket::api
aws:iam/rolePolicyAttachment:RolePolicyAttachment  urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:iam/rolePolicyAttachment:RolePolicyAttachment::api4c238266
aws:apigateway/restApi:RestApi                     urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:apigateway/restApi:RestApi::api
aws:s3/bucketObject:BucketObject                   urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:s3/bucketObject:BucketObject::api4c238266/index.html
aws:apigateway/deployment:Deployment               urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:apigateway/deployment:Deployment::api
aws:lambda/permission:Permission                   urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:lambda/permission:Permission::api-e7f981a5
aws:apigateway/stage:Stage                         urn:pulumi:staging::kmbal-mono::aws-apigateway:index:RestAPI$aws:apigateway/stage:Stage::api


Found no pending operations associated with kmbal/staging

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/bcheidemann-kmbal
User           bcheidemann-kmbal
Organizations  bcheidemann-kmbal, kmbal
Token type     personal

Pulumi locates its logs in /var/folders/qj/tm_rd4fj0pz80xsbrs3y0zvm0000gn/T/ by default
warning: Failed to get information about the Pulumi program's dependencies: could not find either /Users/benheidemann/repos/kmbal-monorepo/packages/infrastructure/yarn.lock or /Users/benheidemann/repos/kmbal-monorepo/packages/infrastructure/package-lock.json
```

### Additional context

This issue seems to have been introduced in `@pulumi/pulumi` version `3.108.0` (version `3.107.0` works as expected).

Here is the root `package.json` from my workspace:

```json
{
  "name": "[REDACTED]",
  "scripts": {},
  "workspaces": ["packages/*"],
  "dependencies": {
    "@pulumi/aws": "^6.0.0",
    "@pulumi/aws-apigateway": "^2.0.0",
    "@pulumi/awsx": "^2.0.2",
    "@pulumi/pulumi": "^3.113.0",
    "typescript": "^5.0.0"
  },
  "devDependencies": {
    "@types/folder-hash": "^4.0.4",
    "@types/node": "^18",
    "dotenv-cli": "^7.3.0",
    "folder-hash": "^4.0.4"
  }
}
```

And here is a working version:

```json
{
  "name": "kmbal-monorepo",
  "scripts": {},
  "workspaces": ["packages/*"],
  "dependencies": {
    "@pulumi/aws": "6.18.2",
    "@pulumi/aws-apigateway": "2.1.1",
    "@pulumi/awsx": "2.4.0",
    "@pulumi/pulumi": "3.107.0",
    "typescript": "5.4.5"
  },
  "devDependencies": {
    "@types/folder-hash": "^4.0.4",
    "@types/node": "^18",
    "dotenv-cli": "^7.3.0",
    "folder-hash": "^4.0.4"
  }
}
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20240621--sdk-nodejs--avoid-an-unhandled-error-when-dependencies-is-missing-from-package-json-during-closure-serialization.yaml
+++ b/changelog/pending/20240621--sdk-nodejs--avoid-an-unhandled-error-when-dependencies-is-missing-from-package-json-during-closure-serialization.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/nodejs
+  description: Avoid an unhandled error when `dependencies` is missing from `package.json` during closure serialization

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -257,8 +257,10 @@ async function allFoldersForPackages(
     // and should not be uploaded.
     const referencedPackages = new Set<string>(includedPackages);
     const packageJSON = computeDependenciesDirectlyFromPackageFile(upath.join(workingDir, "package.json"), logResource);
-    for (const depName of Object.keys(packageJSON.dependencies)) {
-        referencedPackages.add(depName);
+    if (packageJSON.dependencies) {
+        for (const depName of Object.keys(packageJSON.dependencies)) {
+            referencedPackages.add(depName);
+        }
     }
 
     // Find the workspace root, fallback to current working directory if we are not in a workspaces setup.

--- a/tests/integration/integration_nodejs_test.go
+++ b/tests/integration/integration_nodejs_test.go
@@ -1604,6 +1604,15 @@ func TestCodePathsWorkspaceTSC(t *testing.T) {
 	})
 }
 
+//nolint:paralleltest // ProgramTest calls t.Parallel()
+func TestCodePathsNoDependencies(t *testing.T) {
+	integration.ProgramTest(t, &integration.ProgramTestOptions{
+		Dir:          filepath.Join("nodejs", "codepaths-no-dependencies"),
+		Dependencies: []string{"@pulumi/pulumi"},
+		Quick:        true,
+	})
+}
+
 // Test that the resource stopwatch doesn't contain a negative time.
 func TestNoNegativeTimingsOnRefresh(t *testing.T) {
 	if runtime.GOOS == WindowsOS {

--- a/tests/integration/nodejs/codepaths-no-dependencies/Pulumi.yaml
+++ b/tests/integration/nodejs/codepaths-no-dependencies/Pulumi.yaml
@@ -1,0 +1,3 @@
+name: codepaths-no-dependencies
+runtime: nodejs
+description: Basic test case for codepaths with no dependencies in package.json

--- a/tests/integration/nodejs/codepaths-no-dependencies/index.ts
+++ b/tests/integration/nodejs/codepaths-no-dependencies/index.ts
@@ -1,0 +1,26 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as runtime from "@pulumi/pulumi/runtime";
+
+(async function() {
+    const deps = await runtime.computeCodePaths();
+
+    const actual = JSON.stringify([...deps.keys()]);
+    const expected = "[]";
+
+    if (actual !== expected) {
+        throw new Error(`Got '${actual}' expected '${expected}'`);
+    }
+})()

--- a/tests/integration/nodejs/codepaths-no-dependencies/package.json
+++ b/tests/integration/nodejs/codepaths-no-dependencies/package.json
@@ -1,0 +1,5 @@
+{
+    "name": "test-codepaths-no-dependencies",
+    "description": "Basic test case for codepaths where there is no dependencies in package.json",
+    "main": "index.ts"
+}

--- a/tests/integration/nodejs/codepaths-no-dependencies/tsconfig.json
+++ b/tests/integration/nodejs/codepaths-no-dependencies/tsconfig.json
@@ -1,0 +1,16 @@
+{
+  "compilerOptions": {
+    "strict": true,
+    "outDir": "bin",
+    "target": "es2016",
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "sourceMap": true,
+    "experimentalDecorators": true,
+    "pretty": true,
+    "noFallthroughCasesInSwitch": true,
+    "noImplicitReturns": true,
+    "forceConsistentCasingInFileNames": true
+  },
+  "files": ["index.ts"]
+}

